src/types/timestamp.rs

```diff
@@ -845,7 +837,52 @@ struct TimestampBuilder {
     nanoseconds: Option<u32>,
 }
 
-impl TimestampBuilder {
+// The marker traits and the state structs are pub in this module, but they do not appear as types
+// in the documentation, they cannot be imported, and they are not nameable from outside this crate.
+
+/// Marker trait indicating a valid state to build a timestamp
+pub trait Buildable {}
+/// Marker trait indicating that the state includes a time, and can be built with an offset.
+pub trait BuildableWithOffset: Buildable {}
+
+macro_rules! state_with_markers {
+    ($name:ident $(,$trait_name:ident)*)=> {
+        #[derive(Debug, Clone, Default)]
+        pub struct $name;
+        $(
+        impl $trait_name for $name {}
+        )*
+    };
+}
+state_with_markers!(Init);
+state_with_markers!(HasYear, Buildable);
+state_with_markers!(HasMonth, Buildable);
+state_with_markers!(HasDay, Buildable);
+state_with_markers!(HasHour);
+state_with_markers!(HasMinute, Buildable, BuildableWithOffset);
+state_with_markers!(HasSeconds, Buildable, BuildableWithOffset);
+state_with_markers!(HasFractionalSeconds, Buildable, BuildableWithOffset);
+
+impl<T> TimestampBuilder<T> {
+    fn change_state<U>(self) -> TimestampBuilder<U> {
+        // TODO: This entire function could be replaced with one line of unsafe code, but is it worthwhile?
+        // unsafe { std::mem::transmute(self) }
```

Comment:
> However, I don't know if it actually matters when compiling in release mode because when I tried that on godbolt, it optimized away all of my sample code. 🤷‍♂️

Relatable! 😅

I think the `--release` mode is likely to be able to see through this and turn it into something very similar to `transmute`. However, even if it's not, this is something that we should wait to see show up in a perf profile before we reach for `unsafe`.